### PR TITLE
add metadata to ngsild payload measure + doc 

### PR DESCRIPTION
**Pull request by AlvaroVega**
Continuation of PR https://github.com/telefonicaid/iotagent-json/pull/779 to add `metadata` for `observedAt`, `createdAt`  and any other stuff found as ngsild attribute properties.

### CODE REVIEW

File: lib/commonBindings.js [hunk 1]
```diff
@@ -133,6 +133,15 @@ function extractAttributes(device, current, payloadType) {
                                     ent.value = entity[k].object;
                                 }
                             }
+                            // Add other stuff as metadata
+                            for (let key in entity[k]) {
+                                if (!['type', 'value', 'object'].includes(key.toLowerCase())) {
```
Comment (fgalan) [inline lib/commonBindings.js:138]: This comes from NGSIv2 allowed fields for attributes? In this case, note `object` is not one of these (only `value`, `type` and `metadata` are allowed, the later one doesn't apply here).

Comment (AlvaroVega) [reply lib/commonBindings.js:138]: object in ngsild is transformed in to value for ngsiv2, but in this point we are consulting ngsild entity, which has object key.

Comment (fgalan) [reply lib/commonBindings.js:138]: So you mean that in NGSI-LD both `value` and `object` are used for attribute value as shown in the following examples?

```
    "status": {
        "type": "Property",
        "value": "free",
        "observedAt": "2018-09-21T12:00:00Z"
    },
```

```
    "refParkingSite": {
        "type": "Relationship",
        "object": "urn:ngsi-ld:ParkingSite:santander:daoiz_velarde_1_5"
    },
```

File: test/unit/ngsiv2/contextRequests/ngsildPayloadMeasure2.json [hunk 2]
```diff
@@ -3,7 +3,10 @@
     "type": "AnMQTTDevice",
     "status": {
         "type": "Property",
-        "value": "free"
+        "value": "free",
+        "metadata": {
+            "observedAt":"2012-09-21T12:00:00Z"
+        }
```
Comment (fgalan) [inline test/unit/ngsiv2/contextRequests/ngsildPayloadMeasure2.json:9]: This file is for the mock expecations at CB?

This syntax is not correct for metadata. The values for `metadata` sub-keys are objects with `value` and `type`, so it should be:

```suggestion
        "metadata": {
            "observedAt":
                "value": "2012-09-21T12:00:00Z",
                "type": ??
        }
```

What I'm not sure if which `type` to use here... Some candidates: `ngsild`, `ngsildMetadata`, `ngsildTheNameTheyUseForKindOfThingsInTheirSpec`.

File: lib/commonBindings.js (hunk 1 above)
Comment (AlvaroVega) [reply lib/commonBindings.js:138]: Yes, these keys are already handled and should not added to metadata as other stuff.

Comment (fgalan) [reply lib/commonBindings.js:138]: NTC

File: test/unit/ngsiv2/contextRequests/ngsildPayloadMeasure2.json (hunk 2 above)
Comment (mapedraza) [reply test/unit/ngsiv2/contextRequests/ngsildPayloadMeasure2.json:9]: `propertyOfProperty`?

Comment (fgalan) [reply test/unit/ngsiv2/contextRequests/ngsildPayloadMeasure2.json:9]: Another possibility is that the JSON type of the value decides, applying the same rules than in CB: https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/orion-api.md#partial-representations

Comment (AlvaroVega) [reply test/unit/ngsiv2/contextRequests/ngsildPayloadMeasure2.json]: What about an empty 'type'? CB will fill an empty type with a proper type ?

File: docs/usermanual.md [hunk 3]
```diff
@@ -243,7 +243,11 @@ Example of these `ngsild` payloads are the following ones:
 }
 ```
 
-Note that array of entities are handled as a multiple measure, each entity is a measure.
+Some additional considerations to take into account:
+
+-   In the case of array of entities, they are handled as a multiple measure, i.e. each entity is a measure.
+-   The `type` of the attribute is the one used in the provision of the attribute, not the one in the measure. The execption is the autoprovisioned devices case, in which case the `type` of the attribute is taken from the measure (given the attribte lacks proviosioned type). In this latter case, if the attribute `type` is not included in the measure the [explicit type omission rules for Context Broker](https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/orion-api.md#partial-representations) are also taken into account in this case.
```
Comment (fgalan) [inline docs/usermanual.md:249]:
```suggestion
-   The `type` of the attribute is the one used in the provision of the attribute, not the one in the measure. The exception is the autoprovisioned devices case, in which case the `type` of the attribute is taken from the measure (given the attribute lacks proviosioned type). In this latter case, if the attribute `type` is not included in the measure the [explicit type omission rules for Context Broker](https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/orion-api.md#partial-representations) are also taken into account in this case.
```

File: docs/usermanual.md [hunk 4]
```diff
@@ -243,7 +243,11 @@ Example of these `ngsild` payloads are the following ones:
 }
 ```
 
-Note that array of entities are handled as a multiple measure, each entity is a measure.
+Some additional considerations to take into account:
+
+-   In the case of array of entities, they are handled as a multiple measure, i.e. each entity is a measure.
+-   The `type` of the attribute is the one used in the provision of the attribute, not the one in the measure. The exception is the autoprovisioned devices case, in which case the `type` of the attribute is taken from the measure (given the attribute lacks proviosioned type). In this latter case, if the attribute `type` is not included in the measure the [explicit type omission rules for Context Broker](https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/orion-api.md#partial-representations) are also taken into account in this case.
+-   In the case of NGSI-LD, fields different from `type` or `value` (e.g. `observedAt` in the examples above) are include as NGSI-v2 metadata in the entity corresponding to the measure at Context Broker.
```
Comment (fgalan) [inline docs/usermanual.md:250]:
```suggestion
-   In the case of NGSI-LD, fields different from `type`, `value` or `object` (e.g. `observedAt` in the examples above) are include as NGSI-v2 metadata in the entity corresponding to the measure at Context Broker.
```

Comment (fgalan) [reply docs/usermanual.md:250]: @AlvaroVega pls check my suggestion is correct (not fully sure...)

File: docs/usermanual.md [hunk 5]
```diff
@@ -243,7 +243,11 @@ Example of these `ngsild` payloads are the following ones:
 }
 ```
 
-Note that array of entities are handled as a multiple measure, each entity is a measure.
+Some additional considerations to take into account:
+
+-   In the case of array of entities, they are handled as a multiple measure, i.e. each entity is a measure.
+-   The `type` of the attribute is the one used in the provision of the attribute, not the one in the measure. The exception is the autoprovisioned devices case, in which case the `type` of the attribute is taken from the measure (given the attribute lacks proviosioned type). In this latter case, if the attribute `type` is not included in the measure the [explicit type omission rules for Context Broker](https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/orion-api.md#partial-representations) are also taken into account in this case.
+-   In the case of NGSI-LD, fields different from `type`, `value` or `object` (e.g. `observedAt` in the examples above) are include as NGSI-v2 metadata in the entity corresponding to the measure at Context Broker.
```
Comment (fgalan) [inline docs/usermanual.md:250]:
```suggestion
-   In the case of NGSI-LD, fields different from `type`, `value` or `object` (e.g. `observedAt` in the examples above) are include as NGSI-v2 metadata in the entity corresponding to the measure at Context Broker. Note IOTA doesn't provide the `type` for that metadata, so the Context Broker applies [a default type based in the metadat `value` JSON type](https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/orion-api.md#partial-representations).
```

File: docs/usermanual.md [hunk 6]
```diff
@@ -243,7 +243,11 @@ Example of these `ngsild` payloads are the following ones:
...
+Some additional considerations to take into account:
+
+-   In the case of array of entities, they are handled as a multiple measure, i.e. each entity is a measure.
+-   The `type` of the attribute is the one used in the provision of the attribute, not the one in the measure. The exception is the autoprovisioned devices case, in which case the `type` of the attribute is taken from the measure (given the attribute lacks proviosioned type). In this latter case, if the attribute `type` is not included in the measure the [explicit type omission rules for Context Broker](https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/orion-api.md#partial-representations) are also taken into account in this case.
+-   In the case of NGSI-LD, fields different from `type`, `value` or `object` (e.g. `observedAt` in the examples above) are include as NGSI-v2 metadata in the entity corresponding to the measure at Context Broker. Note IOTA doesn't provide the `type` for that metadata, so the Context Broker applies [a default type based in the metadat `value` JSON type](https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/orion-api.md#partial-representations).
```
Comment (fgalan) [inline docs/usermanual.md:250]:
```suggestion
-   In the case of NGSI-LD, fields different from `type`, `value` or `object` (e.g. `observedAt` in the examples above) are include as NGSI-v2 metadata in the entity corresponding to the measure at Context Broker. Note IOTA doesn't provide the `type` for that metadata, so the Context Broker applies [a default type based in the metadata `value` JSON type](https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/orion-api.md#partial-representations).
```

**Pull request merged**